File: server/llm_service.py
"""
Multi-Provider LLM Service - Supports Gemini, OpenAI, and Anthropic
"""
import os
from typing import Optional, List, Dict, Any
from dataclasses import dataclass
from google import genai
from google.genai import types
from openai import OpenAI
from anthropic import Anthropic


# Available models per provider (updated Dec 2025)
AVAILABLE_MODELS = {
    "google": ["gemini-3.0-flash", "gemini-3.0-pro", "gemini-2.5-flash"],
    "openai": ["gpt-5.2", "gpt-5-mini", "gpt-4o"],
    "anthropic": ["claude-opus-4-5-20251101", "claude-sonnet-4-5-20250929", "claude-haiku-4-5-20251101"]
}


@dataclass
class LLMConfig:
    """Configuration for LLM requests"""
    provider: str
    model: str
    temperature: float = 0.2
    max_tokens: int = 4096


class LLMService:
    """Multi-provider LLM service"""

    def __init__(self):
        self._google_client = None
        self._openai_client = None
        self._anthropic_client = None

    def _get_google_client(self) -> genai.Client:
        """Get or create Google GenAI client"""
        if self._google_client is None:
            api_key = os.getenv("GEMINI_API_KEY")
            if not api_key:
                raise ValueError("GEMINI_API_KEY not found in environment")
            self._google_client = genai.Client(api_key=api_key)
        return self._google_client

    def _get_openai_client(self) -> OpenAI:
        """Get or create OpenAI client"""
        if self._openai_client is None:
            api_key = os.getenv("OPENAI_API_KEY")
            if not api_key:
                raise ValueError("OPENAI_API_KEY not found in environment")
            self._openai_client = OpenAI(api_key=api_key)
        return self._openai_client

    def _get_anthropic_client(self) -> Anthropic:
        """Get or create Anthropic client"""
        if self._anthropic_client is None:
            api_key = os.getenv("ANTHROPIC_API_KEY")
            if not api_key:
                raise ValueError("ANTHROPIC_API_KEY not found in environment")
            self._anthropic_client = Anthropic(api_key=api_key)
        return self._anthropic_client

    def get_available_providers(self) -> Dict[str, List[str]]:
        """Get available providers and their models based on configured API keys"""
        available = {}

        if os.getenv("GEMINI_API_KEY"):
            available["google"] = AVAILABLE_MODELS["google"]

        if os.getenv("OPENAI_API_KEY"):
            available["openai"] = AVAILABLE_MODELS["openai"]

        if os.getenv("ANTHROPIC_API_KEY"):
            available["anthropic"] = AVAILABLE_MODELS["anthropic"]

        return available

    async def generate(
        self,
        prompt: str,
        system_prompt: str,
        config: LLMConfig,
        images: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """
        Generate a response from the LLM

        Args:
            prompt: User prompt
            system_prompt: System instructions
            config: LLM configuration
            images: Optional list of images [{mime_type, data}]

        Returns:
            Generated text response
        """
        if config.provider == "google":
            return await self._generate_google(prompt, system_prompt, config, images)
        elif config.provider == "openai":
            return await self._generate_openai(prompt, system_prompt, config, images)
        elif config.provider == "anthropic":
            return await self._generate_anthropic(prompt, system_prompt, config, images)
        else:
            raise ValueError(f"Unknown provider: {config.provider}")

    async def _generate_google(
        self,
        prompt: str,
        system_prompt: str,
        config: LLMConfig,
        images: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """Generate using Google Gemini"""
        client = self._get_google_client()

        # Build content parts
        contents = [prompt]

        if images:
            import base64
            for img in images:
                contents.append(types.Part.from_bytes(
                    data=base64.b64decode(img["data"]),
                    mime_type=img.get("mime_type", "image/png")
                ))

        response = client.models.generate_content(
            model=config.model,
            contents=contents,
            config=types.GenerateContentConfig(
                system_instruction=system_prompt,
                temperature=config.temperature,
                max_output_tokens=config.max_tokens
            )
        )

        return response.text

    async def _generate_openai(
        self,
        prompt: str,
        system_prompt: str,
        config: LLMConfig,
        images: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """Generate using OpenAI"""
        client = self._get_openai_client()

        messages = [
            {"role": "system", "content": system_prompt}
        ]

        # Build user message content
        if images:
            content = [{"type": "text", "text": prompt}]
            for img in images:
                content.append({
                    "type": "image_url",
                    "image_url": {
                        "url": f"data:{img.get('mime_type', 'image/png')};base64,{img['data']}"
                    }
                })
            messages.append({"role": "user", "content": content})
        else:
            messages.append({"role": "user", "content": prompt})

        response = client.chat.completions.create(
            model=config.model,
            messages=messages,
            temperature=config.temperature,
            max_tokens=config.max_tokens
        )

        return response.choices[0].message.content

    async def _generate_anthropic(
        self,
        prompt: str,
        system_prompt: str,
        config: LLMConfig,
        images: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """Generate using Anthropic Claude"""
        client = self._get_anthropic_client()

        # Build user message content
        if images:
            content = []
            for img in images:
                content.append({
                    "type": "image",
                    "source": {
                        "type": "base64",
                        "media_type": img.get("mime_type", "image/png"),
                        "data": img["data"]
                    }
                })
            content.append({"type": "text", "text": prompt})
        else:
            content = prompt

        response = client.messages.create(
            model=config.model,
            max_tokens=config.max_tokens,
            system=system_prompt,
            messages=[{"role": "user", "content": content}],
            temperature=config.temperature
        )

        return response.content[0].text

    async def chat(
        self,
        message: str,
        history: List[Dict[str, str]],
        system_prompt: str,
        config: LLMConfig,
        images: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """
        Chat with conversation history

        Args:
            message: Current user message
            history: Previous messages [{"role": "user"|"assistant", "content": "..."}]
            system_prompt: System instructions
            config: LLM configuration
            images: Optional images for current message

        Returns:
            Assistant response
        """
        if config.provider == "google":
            return await self._chat_google(message, history, system_prompt, config, images)
        elif config.provider == "openai":
            return await self._chat_openai(message, history, system_prompt, config, images)
        elif config.provider == "anthropic":
            return await self._chat_anthropic(message, history, system_prompt, config, images)
        else:
            raise ValueError(f"Unknown provider: {config.provider}")

    async def _chat_google(
        self,
        message: str,
        history: List[Dict[str, str]],
        system_prompt: str,
        config: LLMConfig,
        images: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """Chat using Google Gemini"""
        client = self._get_google_client()

        # Convert history to Gemini format
        contents = []
        for msg in history:
            role = "user" if msg["role"] == "user" else "model"
            contents.append(types.Content(
                role=role,
                parts=[types.Part.from_text(msg["content"])]
            ))

        # Add current message with images
        current_parts = [types.Part.from_text(message)]
        if images:
            import base64
            for img in images:
                current_parts.append(types.Part.from_bytes(
                    data=base64.b64decode(img["data"]),
                    mime_type=img.get("mime_type", "image/png")
                ))

        contents.append(types.Content(
            role="user",
            parts=current_parts
        ))

        response = client.models.generate_content(
            model=config.model,
            contents=contents,
            config=types.GenerateContentConfig(
                system_instruction=system_prompt,
                temperature=config.temperature,
                max_output_tokens=config.max_tokens
            )
        )

        return response.text

    async def _chat_openai(
        self,
        message: str,
        history: List[Dict[str, str]],
        system_prompt: str,
        config: LLMConfig,
        images: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """Chat using OpenAI"""
        client = self._get_openai_client()

        messages = [{"role": "system", "content": system_prompt}]

        # Add history
        for msg in history:
            messages.append({
                "role": msg["role"],
                "content": msg["content"]
            })

        # Add current message with images
        if images:
            content = [{"type": "text", "text": message}]
            for img in images:
                content.append({
                    "type": "image_url",
                    "image_url": {
                        "url": f"data:{img.get('mime_type', 'image/png')};base64,{img['data']}"
                    }
                })
            messages.append({"role": "user", "content": content})
        else:
            messages.append({"role": "user", "content": message})

        response = client.chat.completions.create(
            model=config.model,
            messages=messages,
            temperature=config.temperature,
            max_tokens=config.max_tokens
        )

        return response.choices[0].message.content

    async def _chat_anthropic(
        self,
        message: str,
        history: List[Dict[str, str]],
        system_prompt: str,
        config: LLMConfig,
        images: Optional[List[Dict[str, str]]] = None
    ) -> str:
        """Chat using Anthropic Claude"""
        client = self._get_anthropic_client()

        messages = []

        # Add history
        for msg in history:
            messages.append({
                "role": msg["role"],
                "content": msg["content"]
            })

        # Add current message with images
        if images:
            content = []
            for img in images:
                content.append({
                    "type": "image",
                    "source": {
                        "type": "base64",
                        "media_type": img.get("mime_type", "image/png"),
                        "data": img["data"]
                    }
                })
            content.append({"type": "text", "text": message})
            messages.append({"role": "user", "content": content})
        else:
            messages.append({"role": "user", "content": message})

        response = client.messages.create(
            model=config.model,
            max_tokens=config.max_tokens,
            system=system_prompt,
            messages=messages,
            temperature=config.temperature
        )

        return response.content[0].text


# Global instance
llm_service = LLMService()

File: node-server/src/llm/llm-service.ts
/**
 * Multi-Provider LLM Service
 *
 * Supports Google Gemini, OpenAI, and Anthropic.
 * Node.js port of the Python LLMService.
 */

import { GoogleGenAI } from '@google/genai';
import OpenAI from 'openai';
import Anthropic from '@anthropic-ai/sdk';
import {
  LLMConfig,
  LLMProvider,
  ImageInput,
  ChatMessage,
  GenerateOptions,
  ChatOptions,
  GenerateStructuredOptions,
  AVAILABLE_MODELS,
  DEFAULT_MAX_TOKENS,
  DEFAULT_TEMPERATURE,
} from './types';

export class LLMService {
  private googleClient: GoogleGenAI | null = null;
  private openaiClient: OpenAI | null = null;
  private anthropicClient: Anthropic | null = null;

  /**
   * Get or create Google GenAI client
   */
  private getGoogleClient(apiKey?: string): GoogleGenAI {
    if (apiKey) {
      // Create new client with provided key (don't cache)
      return new GoogleGenAI({ apiKey });
    }

    if (!this.googleClient) {
      const envKey = process.env.GEMINI_API_KEY;
      if (!envKey) {
        throw new Error('GEMINI_API_KEY not found. Configure it in settings or server environment.');
      }
      this.googleClient = new GoogleGenAI({ apiKey: envKey });
    }
    return this.googleClient;
  }

  /**
   * Get or create OpenAI client
   */
  private getOpenaiClient(apiKey?: string): OpenAI {
    if (apiKey) {
      return new OpenAI({ apiKey });
    }

    if (!this.openaiClient) {
      const envKey = process.env.OPENAI_API_KEY;
      if (!envKey) {
        throw new Error('OPENAI_API_KEY not found. Configure it in settings or server environment.');
      }
      this.openaiClient = new OpenAI({ apiKey: envKey });
    }
    return this.openaiClient;
  }

  /**
   * Get or create Anthropic client
   */
  private getAnthropicClient(apiKey?: string): Anthropic {
    if (apiKey) {
      return new Anthropic({ apiKey });
    }

    if (!this.anthropicClient) {
      const envKey = process.env.ANTHROPIC_API_KEY;
      if (!envKey) {
        throw new Error('ANTHROPIC_API_KEY not found. Configure it in settings or server environment.');
      }
      this.anthropicClient = new Anthropic({ apiKey: envKey });
    }
    return this.anthropicClient;
  }

  /**
   * Get available providers based on configured API keys
   */
  getAvailableProviders(): Partial<Record<LLMProvider, string[]>> {
    const available: Partial<Record<LLMProvider, string[]>> = {};

    if (process.env.GEMINI_API_KEY) {
      available.google = AVAILABLE_MODELS.google;
    }
    if (process.env.OPENAI_API_KEY) {
      available.openai = AVAILABLE_MODELS.openai;
    }
    if (process.env.ANTHROPIC_API_KEY) {
      available.anthropic = AVAILABLE_MODELS.anthropic;
    }

    return available;
  }

  /**
   * Normalize config with defaults
   */
  normalizeConfig(config: LLMConfig): Required<Omit<LLMConfig, 'apiKey'>> & { apiKey?: string } {
    return {
      provider: config.provider,
      model: config.model,
      temperature: config.temperature ?? DEFAULT_TEMPERATURE,
      maxTokens: config.maxTokens ?? DEFAULT_MAX_TOKENS,
      apiKey: config.apiKey,
    };
  }

  /**
   * Repair common JSON formatting issues
   */
  repairJson(s: string): string {
    let result = s;
    // Replace : 'value' with : "value"
    result = result.replace(/:\s*'([^']*)'/g, ': "$1"');
    // Remove trailing commas before } or ]
    result = result.replace(/,(\s*[}\]])/g, '$1');
    return result;
  }

  /**
   * Parse JSON with automatic repair on failure
   */
  private parseJsonWithRepair(s: string): unknown {
    try {
      return JSON.parse(s);
    } catch {
      try {
        const repaired = this.repairJson(s);
        return JSON.parse(repaired);
      } catch {
        throw new Error('Failed to parse JSON');
      }
    }
  }

  /**
   * Extract JSON from a response that may contain markdown code blocks
   */
  extractJsonFromResponse(response: string): unknown {
    // Try to find JSON in code block first
    const codeBlockMatch = response.match(/```(?:json)?\s*([\s\S]*?)\s*```/);
    if (codeBlockMatch) {
      return this.parseJsonWithRepair(codeBlockMatch[1]);
    }

    // Try to find raw JSON object
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return this.parseJsonWithRepair(jsonMatch[0]);
    }

    throw new Error('No JSON found in response');
  }

  /**
   * Generate response from Google Gemini
   */
  private async generateGoogle(
    prompt: string,
    systemPrompt: string,
    config: ReturnType<typeof this.normalizeConfig>,
    images?: ImageInput[]
  ): Promise<string> {
    const client = this.getGoogleClient(config.apiKey);

    // Build content parts
    const contents: Array<string | { inlineData: { mimeType: string; data: string } }> = [prompt];

    if (images) {
      for (const img of images) {
        contents.push({
          inlineData: {
            mimeType: img.mimeType || 'image/png',
            data: img.data,
          },
        });
      }
    }

    const response = await client.models.generateContent({
      model: config.model,
      contents,
      config: {
        systemInstruction: systemPrompt,
        temperature: config.temperature,
        maxOutputTokens: config.maxTokens,
      },
    });

    return response.text || '';
  }

  /**
   * Generate response from OpenAI
   */
  private async generateOpenai(
    prompt: string,
    systemPrompt: string,
    config: ReturnType<typeof this.normalizeConfig>,
    images?: ImageInput[]
  ): Promise<string> {
    const client = this.getOpenaiClient(config.apiKey);

    const messages: OpenAI.ChatCompletionMessageParam[] = [
      { role: 'system', content: systemPrompt },
    ];

    // Build user message content
    if (images && images.length > 0) {
      const content: OpenAI.ChatCompletionContentPart[] = [
        { type: 'text', text: prompt },
      ];
      for (const img of images) {
        content.push({
          type: 'image_url',
          image_url: {
            url: `data:${img.mimeType || 'image/png'};base64,${img.data}`,
          },
        });
      }
      messages.push({ role: 'user', content });
    } else {
      messages.push({ role: 'user', content: prompt });
    }

    const response = await client.chat.completions.create({
      model: config.model,
      messages,
      temperature: config.temperature,
      max_completion_tokens: config.maxTokens,
    });

    return response.choices[0]?.message?.content || '';
  }

  /**
   * Generate response from Anthropic Claude
   */
  private async generateAnthropic(
    prompt: string,
    systemPrompt: string,
    config: ReturnType<typeof this.normalizeConfig>,
    images?: ImageInput[]
  ): Promise<string> {
    const client = this.getAnthropicClient(config.apiKey);

    // Build user message content
    let content: Anthropic.MessageCreateParams['messages'][0]['content'];

    if (images && images.length > 0) {
      const parts: Anthropic.ContentBlockParam[] = [];
      for (const img of images) {
        parts.push({
          type: 'image',
          source: {
            type: 'base64',
            media_type: (img.mimeType || 'image/png') as 'image/png' | 'image/gif' | 'image/webp' | 'image/jpeg',
            data: img.data,
          },
        });
      }
      parts.push({ type: 'text', text: prompt });
      content = parts;
    } else {
      content = prompt;
    }

    const response = await client.messages.create({
      model: config.model,
      max_tokens: config.maxTokens,
      system: systemPrompt,
      messages: [{ role: 'user', content }],
      temperature: config.temperature,
    });

    const textBlock = response.content.find(c => c.type === 'text');
    return textBlock && 'text' in textBlock ? textBlock.text : '';
  }

  /**
   * Generate a response from the LLM
   */
  async generate(options: GenerateOptions): Promise<string> {
    const { prompt, systemPrompt, config, images } = options;
    const normalizedConfig = this.normalizeConfig(config);

    switch (config.provider) {
      case 'google':
        return this.generateGoogle(prompt, systemPrompt, normalizedConfig, images);
      case 'openai':
        return this.generateOpenai(prompt, systemPrompt, normalizedConfig, images);
      case 'anthropic':
        return this.generateAnthropic(prompt, systemPrompt, normalizedConfig, images);
      default:
        throw new Error(`Unknown provider: ${config.provider}`);
    }
  }

  /**
   * Generate structured JSON response
   */
  async generateStructured(options: GenerateStructuredOptions): Promise<unknown> {
    const { prompt, systemPrompt, config } = options;
    const normalizedConfig = this.normalizeConfig(config);

    switch (config.provider) {
      case 'google':
        return this.generateStructuredGoogle(prompt, systemPrompt, normalizedConfig);
      case 'openai':
        return this.generateStructuredOpenai(prompt, systemPrompt, normalizedConfig);
      case 'anthropic':
        return this.generateStructuredAnthropic(prompt, systemPrompt, normalizedConfig);
      default:
        throw new Error(`Unknown provider: ${config.provider}`);
    }
  }

  private async generateStructuredGoogle(
    prompt: string,
    systemPrompt: string,
    config: ReturnType<typeof this.normalizeConfig>
  ): Promise<unknown> {
    const client = this.getGoogleClient(config.apiKey);

    const response = await client.models.generateContent({
      model: config.model,
      contents: [prompt],
      config: {
        systemInstruction: systemPrompt,
        temperature: config.temperature,
        maxOutputTokens: config.maxTokens,
        responseMimeType: 'application/json',
      },
    });

    return this.extractJsonFromResponse(response.text || '');
  }

  private async generateStructuredOpenai(
    prompt: string,
    systemPrompt: string,
    config: ReturnType<typeof this.normalizeConfig>
  ): Promise<unknown> {
    const client = this.getOpenaiClient(config.apiKey);

    const response = await client.chat.completions.create({
      model: config.model,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      temperature: config.temperature,
      max_completion_tokens: config.maxTokens,
      response_format: { type: 'json_object' },
    });

    const content = response.choices[0]?.message?.content || '{}';
    return JSON.parse(content);
  }

  private async generateStructuredAnthropic(
    prompt: string,
    systemPrompt: string,
    config: ReturnType<typeof this.normalizeConfig>
  ): Promise<unknown> {
    const client = this.getAnthropicClient(config.apiKey);

    // Anthropic doesn't have native JSON mode, so we enforce via prompt
    const enhancedSystem = systemPrompt +
      '\n\nIMPORTANT: Respond with ONLY a valid JSON object. No markdown, no explanation outside the JSON.';

    const response = await client.messages.create({
      model: config.model,
      max_tokens: config.maxTokens,
      system: enhancedSystem,
      messages: [{ role: 'user', content: prompt }],
      temperature: config.temperature,
    });

    const textBlock = response.content.find(c => c.type === 'text');
    const text = textBlock && 'text' in textBlock ? textBlock.text : '';
    return this.extractJsonFromResponse(text);
  }

  /**
   * Chat with conversation history
   */
  async chat(options: ChatOptions): Promise<string> {
    const { message, history, systemPrompt, config, images } = options;
    const normalizedConfig = this.normalizeConfig(config);

    switch (config.provider) {
      case 'google':
        return this.chatGoogle(message, history, systemPrompt, normalizedConfig, images);
      case 'openai':
        return this.chatOpenai(message, history, systemPrompt, normalizedConfig, images);
      case 'anthropic':
        return this.chatAnthropic(message, history, systemPrompt, normalizedConfig, images);
      default:
        throw new Error(`Unknown provider: ${config.provider}`);
    }
  }

  private async chatGoogle(
    message: string,
    history: ChatMessage[],
    systemPrompt: string,
    config: ReturnType<typeof this.normalizeConfig>,
    images?: ImageInput[]
  ): Promise<string> {
    const client = this.getGoogleClient(config.apiKey);

    // Convert history to Gemini format
    const contents: Array<{ role: 'user' | 'model'; parts: Array<{ text: string }> }> = [];

    for (const msg of history) {
      contents.push({
        role: msg.role === 'user' ? 'user' : 'model',
        parts: [{ text: msg.content }],
      });
    }

    // Add current message with images
    const currentParts: Array<{ text: string } | { inlineData: { mimeType: string; data: string } }> = [
      { text: message },
    ];

    if (images) {
      for (const img of images) {
        currentParts.push({
          inlineData: {
            mimeType: img.mimeType || 'image/png',
            data: img.data,
          },
        });
      }
    }

    contents.push({
      role: 'user',
      parts: currentParts as Array<{ text: string }>,
    });

    const response = await client.models.generateContent({
      model: config.model,
      contents,
      config: {
        systemInstruction: systemPrompt,
        temperature: config.temperature,
        maxOutputTokens: config.maxTokens,
      },
    });

    return response.text || '';
  }

  private async chatOpenai(
    message: string,
    history: ChatMessage[],
    systemPrompt: string,
    config: ReturnType<typeof this.normalizeConfig>,
    images?: ImageInput[]
  ): Promise<string> {
    const client = this.getOpenaiClient(config.apiKey);

    const messages: OpenAI.ChatCompletionMessageParam[] = [
      { role: 'system', content: systemPrompt },
    ];

    // Add history
    for (const msg of history) {
      messages.push({
        role: msg.role,
        content: msg.content,
      });
    }

    // Add current message with images
    if (images && images.length > 0) {
      const content: OpenAI.ChatCompletionContentPart[] = [
        { type: 'text', text: message },
      ];
      for (const img of images) {
        content.push({
          type: 'image_url',
          image_url: {
            url: `data:${img.mimeType || 'image/png'};base64,${img.data}`,
          },
        });
      }
      messages.push({ role: 'user', content });
    } else {
      messages.push({ role: 'user', content: message });
    }

    const response = await client.chat.completions.create({
      model: config.model,
      messages,
      temperature: config.temperature,
      max_completion_tokens: config.maxTokens,
    });

    return response.choices[0]?.message?.content || '';
  }

  private async chatAnthropic(
    message: string,
    history: ChatMessage[],
    systemPrompt: string,
    config: ReturnType<typeof this.normalizeConfig>,
    images?: ImageInput[]
  ): Promise<string> {
    const client = this.getAnthropicClient(config.apiKey);

    const messages: Anthropic.MessageCreateParams['messages'] = [];

    // Add history
    for (const msg of history) {
      messages.push({
        role: msg.role,
        content: msg.content,
      });
    }

    // Add current message with images
    if (images && images.length > 0) {
      const parts: Anthropic.ContentBlockParam[] = [];
      for (const img of images) {
        parts.push({
          type: 'image',
          source: {
            type: 'base64',
            media_type: (img.mimeType || 'image/png') as 'image/png' | 'image/gif' | 'image/webp' | 'image/jpeg',
            data: img.data,
          },
        });
      }
      parts.push({ type: 'text', text: message });
      messages.push({ role: 'user', content: parts });
    } else {
      messages.push({ role: 'user', content: message });
    }

    const response = await client.messages.create({
      model: config.model,
      max_tokens: config.maxTokens,
      system: systemPrompt,
      messages,
      temperature: config.temperature,
    });

    const textBlock = response.content.find(c => c.type === 'text');
    return textBlock && 'text' in textBlock ? textBlock.text : '';
  }
}

// Global instance
export const llmService = new LLMService();
